### DISCUSSION TITLE
CPU cores not being used. Performance issue.

### DISCUSSION THREAD

**Discussion by gnat** [php/frankenphp#294] [Q&A] [answered]
Hello I'm doing a comparison between caddy + php-fpm and frankenphp. Really excited about this project, but performance isn't anywhere in line with what I was expecting.

Is there something I'm obviously missing or have misconfigured? Seems like frankenphp just isn't utilising the resources available to it.

* index.php
  * `<?php echo "hello";`
* tested using [hey](https://github.com/rakyll/hey)
  * `ulimit -n 1000000` before running hey
  * `./hey -z 10s -c 1000 http://127.0.0.1:8888/index.php`

## caddy + php-fpm
* `ulimit -n 1000000` before running caddy
* :green_circle: ~16,000 r/s. Full saturation of all cpu cores. Great!

**php-fpm.conf**
```
[main]
pm = static
pm.max_children = 1000
pm.max_requests = 1000
```
## frankenphp
* `ulimit -n 1000000` before running frankenphp (caddy)
* :red_circle: ~400 r/s. Cores barely seeing any saturation of cpu, maybe 20% at most. :frowning: 

**Caddyfile**
```
frankenphp {
	num_threads 1000
}
```

Any help appreciated! Thanks for the otherwise awesome project.

**Comment by withinboredom** [comment]
After messing around ... I see RPS soar to anywhere from 400 - 1000 RPS on real server hardware and then crash down to about 100 RPS after a few minutes and unable to recover. I set GOMAXPROCS to 1000 and indeed, see 1000+ threads being utilized but almost 0 CPU. Setting GOMAXPROCS up (much) higher seems to let the RPS get higher and eventually recover, but I wonder what is going on...

<img width="689" alt="image" src="https://github.com/dunglas/frankenphp/assets/1883296/16a264e1-ae15-4b1e-b67b-697fb539d4c3">

<img width="1042" alt="image" src="https://github.com/dunglas/frankenphp/assets/1883296/12494011-2348-4fb1-9ecc-9cd79497e086">

cc: @dunglas

**Reply by withinboredom** [reply]
Here's a pprof profile taken from the server while under "load."
[profile.zip](https://github.com/dunglas/frankenphp/files/13289678/profile.zip)

**Reply by withinboredom** [reply]
Taking a look at the full stack dump, it appears most threads are stuck waiting for `runtime.goexit()`, which may indicate most threads are waiting for some deferred task to complete but it never completes.

**Reply by dunglas** [reply]
Just out of curiosity, are you using a fresh build from today? I merged two patches here that may (or may not) help.

**Reply by withinboredom** [reply]
Hmm, let me check. Just came by to say that it doesn't look like we're calling pthread_exit() (see https://github.com/golang/go/issues/20395#issuecomment-302236796) which might be why all these threads are hanging around.

**Reply by dunglas** [reply]
@withinboredom the thread leak could also explain the memory leak we're tracking!

Could you also take a "goroutine" profile to see if we have dnagling goroutines or a dangling threads created by the C lib?

**Reply by withinboredom** [reply]
[goroutine.txt](https://github.com/dunglas/frankenphp/files/13289956/goroutine.txt)

**Reply by dunglas** [reply]
Thanks!

We shouldn't have to call `pthread_exit()` because we use a thread pool library, threads are never destroyed because they are reused. But yet, the issue is probably somewhere in this part.

**Reply by withinboredom** [reply]
Current version:
```
frankenphp --version
FrankenPHP 9d56f2530a2573d922e5fdf52c1af80c3da20d2b PHP 8.2.12 Caddy v2.7.5 h1:HoysvZkLcN2xJExEepaFHK92Qgs7xAiCFydN5x5Hs6Q=
```

**Reply by withinboredom** [reply]
Ah, yeah, forgot about the thread pool. Hmm. 

I have a theory that seems to make some sense. GOMAXPROCS and frankenphp `threads` option need to be balanced. In other words, if you have 1000 frankenphp threads, they are at most, going to be creating that many Go threads that will get locked to the frankenphp threads. So, if GOMAXPROCS is the same number or smaller, you'll start running into issues.

Still testing that out.

**Reply by dunglas** [reply]
The `syscall, locked to thread` goroutines look normal: https://groups.google.com/g/golang-nuts/c/wxMP7TpNGbY/m/0sFFpilQEgAJ

The `threads` option is not really related to `GOMAXPROCS`. `threads` controls how many C threads to start (the C thread pool that will execute PHP scripts).

@gnat @withinboredom could you try to increase the number of C threads to see if it improves (or degrades) the numbers?

**Reply by withinboredom** [reply]
Threw `hey` at it this time (useful little tool) and got totally different results. I ran `hey` from a nearby node (not exactly on the LAN, but in the same DC and has a direct-ish 1Gbit connection), hitting `beta.getswytch.com` (an empty page unless logged in, but a bunch of stuff going on to generate it).

Man, this thing can go:
<img width="1280" alt="image" src="https://github.com/dunglas/frankenphp/assets/1883296/00698031-12a0-4e73-a1ef-73da0cf1a904">

It's nearly pushing the node to thermal throttle...

Anyway, here's some fancy stats in various conditions. 

## Less procs than threads

Environment:

```
            - name: FRANKENPHP_CONFIG
              value: |
                num_threads 500
            - name: GOMAXPROCS
              value: '250'
            - name: GOGC
              value: "400"
            - name: CADDY_GLOBAL_OPTIONS
              value: debug
            - name: CADDY_ADMIN
              value: ':2019'
```

Results:

```
Summary:
  Total:        60.6352 secs
  Slowest:      3.3463 secs
  Fastest:      0.2423 secs
  Average:      0.8286 secs
  Requests/sec: 1201.1679


Response time histogram:
  0.242 [1]     |
  0.553 [7744]  |■■■■■■■■
  0.863 [39199] |■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■
  1.173 [18205] |■■■■■■■■■■■■■■■■■■■
  1.484 [5510]  |■■■■■■
  1.794 [1562]  |■■
  2.105 [455]   |
  2.415 [113]   |
  2.725 [30]    |
  3.036 [8]     |
  3.346 [6]     |


Latency distribution:
  10% in 0.5459 secs
  25% in 0.6403 secs
  50% in 0.7731 secs
  75% in 0.9559 secs
  90% in 1.1882 secs
  95% in 1.3593 secs
  99% in 1.7391 secs

Details (average, fastest, slowest):
  DNS+dialup:   0.0025 secs, 0.2423 secs, 3.3463 secs
  DNS-lookup:   0.0013 secs, 0.0000 secs, 0.1379 secs
  req write:    0.0000 secs, 0.0000 secs, 0.0109 secs
  resp wait:    0.8260 secs, 0.0525 secs, 3.3462 secs
  resp read:    0.0001 secs, 0.0000 secs, 0.0065 secs

Status code distribution:
  [200] 72833 responses
```

Observations:

Memory usage: 4.9 gb after 8 minutes, 21 seconds
Threads: 737 (total)
CPU Max-ish: 70%

## Equal procs and threads

Environment:

```
            - name: FRANKENPHP_CONFIG
              value: |
                num_threads 500
            - name: GOMAXPROCS
              value: '500'
            - name: GOGC
              value: "400"
            - name: CADDY_GLOBAL_OPTIONS
              value: debug
            - name: CADDY_ADMIN
              value: ':2019'

```

Results:

```
Summary:
  Total:        61.2683 secs
  Slowest:      3.7932 secs
  Fastest:      0.3272 secs
  Average:      0.8195 secs
  Requests/sec: 1212.5851


Response time histogram:
  0.327 [1]     |
  0.674 [24396] |■■■■■■■■■■■■■■■■■■■■■■■■■■
  1.020 [37461] |■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■
  1.367 [9246]  |■■■■■■■■■■
  1.714 [2050]  |■■
  2.060 [461]   |
  2.407 [191]   |
  2.753 [246]   |
  3.100 [169]   |
  3.447 [54]    |
  3.793 [18]    |


Latency distribution:
  10% in 0.5380 secs
  25% in 0.6380 secs
  50% in 0.7604 secs
  75% in 0.9221 secs
  90% in 1.1407 secs
  95% in 1.3172 secs
  99% in 1.9787 secs

Details (average, fastest, slowest):
  DNS+dialup:   0.0037 secs, 0.3272 secs, 3.7932 secs
  DNS-lookup:   0.0024 secs, 0.0000 secs, 0.2272 secs
  req write:    0.0000 secs, 0.0000 secs, 0.0293 secs
  resp wait:    0.8156 secs, 0.0817 secs, 3.7931 secs
  resp read:    0.0001 secs, 0.0000 secs, 0.0047 secs

Status code distribution:
  [200] 74293 responses
```

Memory usage: 4.4 gb after 4 minutes, 56 seconds
Threads: 1000 (total)
Cpu Max-ish: 60%

## More procs than threads

Environment:

```
            - name: FRANKENPHP_CONFIG
              value: |
                num_threads 500
            - name: GOMAXPROCS
              value: '1000'
            - name: GOGC
              value: "400"
            - name: CADDY_GLOBAL_OPTIONS
              value: debug
            - name: CADDY_ADMIN
              value: ':2019'
```

Results:

```
Summary:
  Total:        60.5793 secs
  Slowest:      2.3661 secs
  Fastest:      0.1662 secs
  Average:      0.6860 secs
  Requests/sec: 1451.3377


Response time histogram:
  0.166 [1]     |
  0.386 [632]   |■
  0.606 [36118] |■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■
  0.826 [37128] |■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■
  1.046 [11595] |■■■■■■■■■■■■
  1.266 [2163]  |■■
  1.486 [253]   |
  1.706 [28]    |
  1.926 [2]     |
  2.146 [0]     |
  2.366 [1]     |


Latency distribution:
  10% in 0.5609 secs
  25% in 0.5826 secs
  50% in 0.6842 secs
  75% in 0.7414 secs
  90% in 0.8592 secs
  95% in 0.9571 secs
  99% in 1.1451 secs

Details (average, fastest, slowest):
  DNS+dialup:   0.0021 secs, 0.1662 secs, 2.3661 secs
  DNS-lookup:   0.0012 secs, 0.0000 secs, 0.1530 secs
  req write:    0.0001 secs, 0.0000 secs, 0.0424 secs
  resp wait:    0.6837 secs, 0.0289 secs, 2.3661 secs
  resp read:    0.0001 secs, 0.0000 secs, 0.0126 secs

Status code distribution:
  [200] 87921 responses
```

Observations:

Memory usage: 4.8 gb after 3 minutes, 44 seconds
Threads: 961 (total)
Cpu Max-ish: 45%

## Conclusion

According to [the docs](https://go.dev/src/runtime/cgocall.go), GOMAXPROCS kills these sleeping threads, but this incurs an overhead (aka, more CPU usage), and thus less RPS total, when GOMAXPROCS is less than or equal to the number of C threads. Having more sleeping threads lying around means that more memory will be utilized, but more RPS. So, there is a tradeoff in the number of Go threads + FrankenPHP threads.

Want less CPU usage, and more RPS? Give it more PROCS than frankenphp threads, but it will use slightly more memory. Note that my environment included a GOGC value of 400% (to prevent GC from occurring during this test), and setting it to 50% on the last test reduces the memory usage to 3.6gb and RPS by ~100. So, there are some further tunings to be made.

In the end, it requires tuning for whatever load you're giving it ... we just have to figure out all the knobs. There are probably some on the PHP side as well, I just haven't fiddled with it.

**Comment by gnat** [comment]
> could you try to increase the number of C threads to see if it improves (or degrades) the numbers?

Yup this is the first thing I tried before posting this discussion- it unfortunately has no effect as far as I can tell.

**Comment by gnat** [comment]
Congrats on the big 1.1.0!

Unfortunately still experiencing the exact same issue with the latest release.

* `htop` reports very little CPU usage by `frankenphp`.
* `GOMAXPROCS` seems to do nothing to `frankenphp`.
* `caddy` + `php-fpm` drastically outperforms `frankenphp` by **over 10x** because it's leveraging all cores to the maximum on the CPU. (`htop` confirms full CPU saturation.)

There may be a mis-configuration of `frankenphp` being overlooked by the devs? 

...maybe the runtime config isn't being propagated?

**Reply by withinboredom** [reply]
I cannot reproduce your numbers. I can completely saturate a 1Gb connection using your original post, using about 60% of my CPU. What architecture are you running on, how much bandwidth do you have, etc?

**Reply by gnat** [reply]
How can you saturate a 1Gb connection at 1,000 r/s (..unless your requests are ~1 Mb each)? The numbers seem off by many orders of magnitude.

**Comment by gdaszuta** [comment]
I was wondering what is happening on my app that I was testing in two variants – with or without Symfony Runtime – Performance with runtime (measured as fulfilled requests) was around 12% of that without runtime. The answer was, memory limit in php.ini – with 1G limit, in Worker mode, FrankenPHP was utilizing only 2 cores. After raising this to 16G, response rate in Worker mode in comparison to legacy mode raised to 440%.

So it does seems, that memory limit is global per all threads – which sort of makes sense, I guess?

**Comment by gnat** [comment]
* Can confirm FrankenPHP seems to be only utilising roughly 2 cores.

![image](https://github.com/dunglas/frankenphp/assets/24665/3491e157-8eaf-457a-9e74-680cf66af941)

* Stress test: `ulimit -n 1000000; ./hey -z 10s -c 1000 http://127.0.0.1/index.php`
* index.php: `<?php echo "hello";`
* `memory_limit=20G` did nothing in my case, still identical performance.

![image](https://github.com/dunglas/frankenphp/assets/24665/e5a088c6-de81-40bf-83fe-eead6fced7c2)


Testing in non-worker, normal PHP mode.

**Reply by gdaszuta** [reply]
Maybe it's configuration difference?

I am running something like that:

```$ cat docker-compose.yml 
version: '2'
services:
  runtime:
    build:
      context: .
    ports:
      - 8068:8068
    volumes:
        - .:/app

    environment:
      - SERVER_NAME=http://localhost:8068
      - FRANKENPHP_CONFIG=worker ./public/index.php
      - APP_RUNTIME=Runtime\FrankenPhpSymfony\Runtime

  noruntime:
    build:
      context: .
    ports:
      - 8067:8067
    volumes:
        - .:/app
    environment:
      - SERVER_NAME=http://localhost:8067
```

```$ cat Dockerfile 
FROM dunglas/frankenphp:1.1.0-php8.2

ARG COMPOSER_SHA=76a7060ccb93902cd7576b67264ad91c8a2700e2
WORKDIR /usr/local/bin

RUN apt update \
 && apt install -y git p7zip-full

RUN curl https://raw.githubusercontent.com/composer/getcomposer.org/${COMPOSER_SHA}/web/installer -s | php -- --quiet \
 && chmod +x composer.phar \
 && ln -fs composer.phar composer

RUN curl -sS https://get.symfony.com/cli/installer | bash \
 && mv /root/.symfony5/bin/symfony /usr/local/bin/symfony

RUN apt install -y htop

WORKDIR /app
```

And with symfony demo it was running fine on both containers, and with my production app, runtime mode was painfully slow. Raising memory_limit seems to solve this problem, and introduce some other problems (but they might be not related to frankenphp).

**Reply by withinboredom** [reply]
Have you tried a different load testing tool? I've noticed some issues with Hey + caddy, in general (related to FIN and FIN ACK packets).

**Comment by gnat** [comment]
Note: Vanilla Caddy will utilise as many cores as it needs, so it puzzles me as to why this is an issue with FrankenPHP

**Comment by gnat** [comment]
Until this is resolved, one can simply run multiple copies of frankenphp at the same time to get an order of magnitude higher throughput (each copy of frankenphp roughly doubles performance) until you saturate your CPU. @dunglas 

`./frankenphp run --watch`
can look like:
`for i in {1..20}; do ./frankenphp run --watch & done`

This isn't ideal, as FrankenPHP should be capable of leveraging more CPU's on its own, but is a short term solution.

**Reply by nickchomey** [reply]
Does this run 20 different instances of caddy and frankenphp?

**Comment by dunglas** [comment, marked as answer]
Version 1.2.2 should improve the situation (especially for benchmarks and stress tests) thanks to https://github.com/dunglas/frankenphp/pull/898.

**Reply by AlliBalliBaba** [reply]
When doing some stress tests myself on 1.2.2 I noticed that cgi mode got indeed faster, but it still doesn't utilize all cores.
Here's what I think the issue might be:
- When Cgo calls a C function it passes on a reference to the current go thread so C can then in turn reference that thread when calling a go function.
- When all the php threads are created in C using `pthread`, all threads will have the same reference to the go thread that created them
- When the C threads are calling `go_handle_request()` in cgi mode, they all reference the same go thread, so all work done inside `go_handle_request()` can only happen on 1 thread and therefore 1 CPU core
- worker mode does not have this issue since in worker mode there is another 'ping-pong' happening between go and C and there's again x function calls for x threads. So the go optimisation works as intended
- Does this sound plausible? Just to be clear - I have no idea how Cgo works under the hood

**Reply by withinboredom** [reply]
Interesting, I thought we got that bottleneck out of there. I'll take a deeper look at it this weekend.

**Reply by dunglas** [reply]
@AlliBalliBaba this is unlikely, the Go runtime doesn't work like that. The Go code should run on the C-thread (and if that wasn't the case, each call would run on a thread selected by the Go runtime).

**Reply by AlliBalliBaba** [reply]
Another theory I have is that go doesn't have enough goroutines 'in reserve' in CGI mode. In `worker.go` when the workers are created, there are x goroutines for x workers being created that can handle all the work on the go side. In cgi mode there is only only 1 goroutine calling `ServeHttp`. Caddy won't spawn more goroutines since from Caddy side everything seems to be fine. The bottleneck only happens when we call back from C to go (there are some [known issues](https://github.com/golang/go/issues/19574) with how Cgo and the go scheduler interact)

**Reply by withinboredom** [reply]
That isn't how Go works, but if you are observing that, then there is probably a bottleneck somewhere else (such as in the code doing the benchmark only sending one request at a time, or something). A good way to test this hypothesis would be to set a `sleep(5)` in the php side and do five requests. They should all end around 5s later instead of 25 seconds later. 5s later means there isn't a single-threaded bottleneck while 25s later means there is.

**Reply by AlliBalliBaba** [reply]
That's the weird part, the code is just a PHP file with `echo "hello world";`. While worker mode scales with the amount of threads that are created in C, CGI mode actually handles less requests as we increase the amount of threads.
Results from my quick 'Hello World' benchmark with `wrk` on 20 CPU cores:

- CGI num_threads 1:                                ` ~3.000     requests/second`
- CGI num_threads 2:                                ` ~2.500     requests/second`
- CGI num_threads 40:                               `~2.000     requests/second`
- Worker mode num_threads 40:               `~40.000   requests/second`

That made me suspect that the issue lies with how C and go threads are communicating, but it's very hard to figure out where exactly the bottleneck lies.
This issue is not that pressing though, performance of CGI mode got a lot better with 1.2.2. The only ones still affected are apps that cannot use worker mode and are running on a server with a lot of CPU cores.

**Reply by withinboredom** [reply]
I think you actually may have stumbled upon something interesting.

It looks like the main issue boils down to how we load in CGI mode. Most of the time is being spent in PHP, not Go.

Since we close the file after executing it and delete the file_handle, we recompile the file from scratch, every time. There may also be a segfault in here as well since during compilation PHP updates interned strings (which isn't thread-safe unless you have opcache enabled).

A better approach would be to leave the file open between requests, sharing the file handles among all threads -- I think that could work -- so it doesn't need to recompile. (see fpm: `zend_destroy_file_handle(&file_handle);` which only happens in specific cases).

TL;DR: we are recompiling the entire project on each and every request in CGI mode.

**Reply by nickchomey** [reply]
Apart from the main issue being discussed here, the most recent test results are quite peculiar. 

First, the cgi drop is not at all linear. (perhaps you could also test 5, 10, 15, 20, 30 threads?) 

Also, worker mode seems much slower than cgi mode, which surely should not be the case. With 40 threads, it results in 1000 requests per thread/2000 per core. But cgi with 1 thread can do 3000/thread. Could you try worker mode with 1, 2, 5, 10, 15, 20, 30 threads as well?

**Reply by nickchomey** [reply]
> TL;DR: we are recompiling the entire project on each and every request in CGI mode.

Does this mean it is not making use of Opcache?

**Reply by AlliBalliBaba** [reply]
I think the decay in cgi mode with more threads probably just comes from an increased amount of context switches the CPU has to do.
Interestingly when there is just 1 worker (`FRANKENPHP_CONFIG=worker /app/public/worker.php 1`), the amount of requests being handled is roughly the same in the 'Hello world' situation (~38.000 rps). That's probably because all the work happens on the go side and PHP can hold up with 1 thread.

If the PHP side has to do some actual work, aka:
```
$a = 0;
for ($i=0; $i < 1_000_000; $i++) {
        $a += $i;
}
echo "Hello world";
```
cgi mode and worker mode become roughly as fast. This confirms my suspicion that the PHP threads are working as intended and the bottleneck lies on the go side if PHP does small to moderate amounts of work.

**Reply by nickchomey** [reply]
> I think the decay in cgi mode with more threads probably just comes from an increased amount of context switches the CPU has to do.

How many CPU cores are available to php? Because, again, the decay is not at all linear/sensical. RPS drops from 3000 to 2500 with just 1 extra thread (which surely the cpu has available, so shouldn't suffer from context switching). And then it only loses another 500 over the next 38 threads. 

Again, I think it would be helpful to see results with something like 1, 2, 5, 10, 15, 20, 30, 40 threads - for both CGI and Worker. Or, at the very least, more than just 1, 2, 40. 

Likewise if you could share some data for both when there's actual work to do.

**Reply by withinboredom** [reply]
I just finished testing things with ensuring that the file is kept in memory on each thread and I didn't see a performance increase. I need to figure out how to step through the C code... I know I've done it before, but it has been awhile. I'll have to take a deeper look tomorrow. I need to get up early, so I should head to bed.

> This confirms my suspicion that the PHP threads are working as intended and the bottleneck lies on the go side if PHP does small to moderate amounts of work.

I now suspect this is the case as well. Looking at it more closely, I think I spot some problems. I'll take a gander on the train in the morning.

**Reply by withinboredom** [reply]
I'm still looking into this, but I did find a point of contention in worker mode. If I can factor it out, we likely can increase the speed of worker mode. I can confirm something fishy with cgi-mode; but I haven't found the culprit yet. I'm still trawling through traces.

**Reply by withinboredom** [reply]
Ok. So this affects both equally, but cgi mode more, for some reason.

Essentially creating a new cgo handle produces a global lock, which means no thread can access any handle until the first one is finished creating it. If you have high load, you can end up with a large number of threads just waiting to create handles.

This is a pretty easy fix and will increase performance dramatically. I should have a PR :soon:

**Reply by withinboredom** [reply]
Ok, made some progress! I have a PoC with the following changes:

- lock-free cgo handles
- not destroying the file handle

The results are not as good as I was expecting. Most of the performance gains are in worker mode ... actually, pretty much all of them ... with a gain of about 10%. The more cores/workers there are, the better the perf improvement will be. I will def open a PR for the lock-free cgo handles, and try to upstream it as well. It's super niche, but probably worth it to include in go and it was tricky af to get right.

There appears to be another bottleneck here, and I'm 99% sure it is on the php side after staring at go stack traces for awhile now.

**Reply by dunglas** [reply]
10% improvement is huge! Thanks a lot for working on this.

**Reply by withinboredom** [reply]
Found it...!!! Its really dumb...

A few PRs are on the way.

**Comment by withinboredom** [comment]
This should be addressed in #933 (which makes cgi-mode faster than worker mode) and #934 (which makes worker mode as fast as cgi-mode)
